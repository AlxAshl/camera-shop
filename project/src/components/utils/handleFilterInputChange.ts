import { ActionCreatorWithNonInferrablePayload } from '@reduxjs/toolkit';
import { store } from '../../store/store';

export const handleFilterInputChange = (filterName: string, filter: string[], action: ActionCreatorWithNonInferrablePayload) => {
  const mutableFilter = Object.assign([], filter);
  if(filter.includes(filterName)) {
    const updatedFilter = mutableFilter.filter((entry) => entry !== filterName);
    store.dispatch(action(updatedFilter));
  }
  else {
    mutableFilter.push(filterName);
    store.dispatch(action(mutableFilter));
  }
};
