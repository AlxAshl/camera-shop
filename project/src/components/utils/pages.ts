import { PAGE_LIMIT } from '../../const';

function getPagesCount(cameraCount: number): number {
  return Math.ceil(cameraCount / PAGE_LIMIT);
}

export default getPagesCount;
