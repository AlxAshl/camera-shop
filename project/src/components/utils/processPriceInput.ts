import { ActionCreatorWithNonInferrablePayload } from '@reduxjs/toolkit';
import { store } from '../../store/store';
import { ProductType } from '../../types/product';

export const setMinInput = (input: string, productsArray: ProductType[], action: ActionCreatorWithNonInferrablePayload) => {
  if(productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    store.dispatch(action([String(input)]));
  }
  if(!productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    let newMinPriceInput;
    const newMin = productsArray.find((product) => product.price <= Number(input));
    newMin?.price === undefined
      ? newMinPriceInput = productsArray[productsArray.length - 1].price
      : newMinPriceInput = newMin?.price;
    store.dispatch(action([String(newMinPriceInput)]));
  }
};

export const setMaxInput = (input: string, productsArray: ProductType[], action: ActionCreatorWithNonInferrablePayload) => {
  if(productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    store.dispatch(action([String(input)]));
  }
  if(!productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    let newMaxPriceInput;
    const newMax = productsArray.find((product) => product.price >= Number(input));
    newMax?.price === undefined
      ? newMaxPriceInput = productsArray[productsArray.length - 1].price
      : newMaxPriceInput = newMax?.price;
    store.dispatch(action([String(newMaxPriceInput)]));
  }
};
