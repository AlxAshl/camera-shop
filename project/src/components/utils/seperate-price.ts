export function seperatePrice(priceValue: number) {
  let returnedPrice = String(priceValue);
  if(returnedPrice.length > 6) {
    returnedPrice = `${returnedPrice.slice(0, 1)} ${returnedPrice.slice(1, 4)} ${returnedPrice.slice(-3)}`;
  }
  if (returnedPrice.length > 3) {
    returnedPrice = `${returnedPrice.slice(-0, -3)} ${returnedPrice.slice(-3)}`;
  }
  return returnedPrice;
}
