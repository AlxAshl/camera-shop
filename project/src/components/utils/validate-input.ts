export function validateInput(target: (EventTarget & HTMLInputElement) | (EventTarget & HTMLTextAreaElement)) {

  if (target.parentElement?.parentElement && target.name === 'rating' && (Number(target.value) > 0)) {
    if (!target.parentElement.parentElement.parentElement?.className.includes('is-valid')) {
      target.parentElement.parentElement.parentElement?.classList.add('is-valid');
      target.parentElement.parentElement.parentElement?.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && target.name !== 'review' && typeof(target.value) === 'string') {
    if (!target.parentElement.parentElement.className.includes('is-valid') && target.value.length > 0) {
      target.parentElement.parentElement.className += ' is-valid';
      target.parentElement.parentElement.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && typeof(target.value) === 'string'){
    if(target.value.length === 0) {
      target.parentElement.parentElement.className += ' is-invalid';
      target.parentElement.parentElement.classList.remove('is-valid');
    }
  }

  if (target.parentElement?.parentElement && typeof(target.value) === 'string') {
    if (!target.parentElement.parentElement.className.includes('is-valid') && target.value.length >= 5) {
      target.parentElement.parentElement.className += ' is-valid';
      target.parentElement.parentElement.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && target.name === 'review' && target.value.length < 5) {
    if (!target.parentElement.parentElement.className.includes('is-invalid')) {
      target.parentElement.parentElement.className += ' is-invalid';
      target.parentElement.parentElement.classList.remove('is-valid');
    }
  }

}
