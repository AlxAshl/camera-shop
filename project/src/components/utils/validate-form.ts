import { reviewDataType } from '../../types/review';
import { validateInput } from './validate-input';


export function validateForm(data: reviewDataType) {
  let validationStatus = true;
  Object.entries(data).forEach(([key, value]) =>{
    if (key !== 'review') {
      if(value === '') {
        const failedInput = document.querySelector(`input[name=${key}]`);
        validateInput(failedInput as HTMLInputElement);
        validationStatus = false;
      }
    }
    if(key === 'review') {
      if(String(value).length < 5) {
        const failedInput = document.querySelector(`textarea[name=${key}]`);
        validateInput(failedInput as HTMLInputElement);
      }
    }
    if(key === 'rating') {
      if(value < 1) {
        const failedInput = document.querySelector(`input[name=${key}]`);
        if(failedInput && failedInput.parentElement?.parentElement){
          failedInput.parentElement.parentElement.parentElement?.classList.add('is-invalid');
        }
        validationStatus = false;
      }
    }
  });
  return validationStatus;
}

