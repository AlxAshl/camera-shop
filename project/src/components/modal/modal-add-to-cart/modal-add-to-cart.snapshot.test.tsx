import { basketProductsMock } from '../../../test/test-mocks';
import { renderWithProviders } from '../../../test/utils/render-with-redux';
import ModalAddToCart from './modal-add-to-cart';

test('Render ModalAddToCart', () => {
  const { asFragment } = renderWithProviders(<ModalAddToCart />, {initialState: { BASKET: { basketProducts: basketProductsMock, productId: 1}, FILTERS: {allProducts: basketProductsMock}}});
  expect(asFragment()).toMatchSnapshot();
});
