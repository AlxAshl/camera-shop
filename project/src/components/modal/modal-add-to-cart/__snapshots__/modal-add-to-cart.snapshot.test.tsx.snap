// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ModalAddToCart 1`] = `
<DocumentFragment>
  <div
    class="modal is-active"
  >
    <div
      class="modal__wrapper"
    >
      <div
        class="modal__overlay"
      />
      <div
        class="modal__content"
      >
        <p
          class="title title--h4"
        >
          Добавить товар в корзину
        </p>
        <div
          class="basket-item basket-item--short"
        >
          <div
            class="basket-item__img"
          >
            <picture>
              <source
                srcset="img/content/look-identify.webp, img/content/look-identify@2x.webp"
                type="image/webp"
              />
              <img
                alt="Фотоаппарат «Орлёнок»"
                height="120"
                src="img/content/look-identify.jpg"
                srcset="img/content/look-identify@2x.jpg"
                width="140"
              />
            </picture>
          </div>
          <div
            class="basket-item__description"
          >
            <p
              class="basket-item__title"
            >
              Look Identify1
            </p>
            <ul
              class="basket-item__list"
            >
              <li
                class="basket-item__list-item"
              >
                <span
                  class="basket-item__article"
                >
                  Артикул:
                </span>
                 
                <span
                  class="basket-item__number"
                >
                  LD2352000
                </span>
              </li>
              <li
                class="basket-item__list-item"
              >
                Фотоаппарат
              </li>
              <li
                class="basket-item__list-item"
              >
                Нулевой уровень
              </li>
            </ul>
            <p
              class="basket-item__price"
            >
              <span
                class="visually-hidden"
              >
                Цена:
              </span>
              126 000 ₽
            </p>
          </div>
        </div>
        <div
          class="modal__buttons"
        >
          <button
            class="btn btn--purple modal__btn modal__btn--fit-width"
            data-testid="cart-button-test"
            id="add-to-cart"
            type="button"
          >
            <svg
              aria-hidden="true"
              height="16"
              width="24"
            >
              <use
                xlink:href="#icon-add-basket"
              />
            </svg>
            Добавить в корзину
          </button>
        </div>
        <button
          aria-label="Закрыть попап"
          class="cross-btn"
          data-testid="cross-btn-test"
          type="button"
        >
          <svg
            aria-hidden="true"
            height="10"
            width="10"
          >
            <use
              xlink:href="#icon-close"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
